File: go/binarysearch/153_find_min.go
package binarysearch

// findMin 寻找旋转排序数组中的最小值
func findMin(nums []int) int {

	n := len(nums)

	l, r := 0, n-1

	for l <= r {

		mid := l + (r-l)>>1

		// 如果旋转次数小于n, 那么左半段有序列表的第一个值，大于右边的所有元素
		// 如果旋转次数等于n， 整体是有序列表

		// 抛弃右侧的数组数据
		if nums[mid] < nums[r] {
			r = mid
		} else {
			l = mid + 1
		}
	}
	// 因为上述循环条件使用 l <= r 的时候，最后 l = mid = r 的时候，会执行 l = mid + 1
	// 所以此时返回 nums[l-1] 或者 nums[r], 举例[*, 10, 1, *], nums[l]=10, nums[r]=1
	return nums[r]
}

File: rust/src/binarysearch/_153_find_min.rs


pub fn find_min(nums: Vec<i32>) -> i32 {

    let (mut l, mut r) = (0, (nums.len()-1) as i32);

    while l <= r {

        let mid = l + ((r-l)>>1);

        if nums[mid as usize] < nums[r as usize] {
            // mid 有可能是最小值，不能是r = mid -1 会错过最小值
            r = mid;
        }else{
            l = mid + 1
        }
    }
    nums[r as usize]
}